onnxruntime/core/framework/tunable.h

```diff
@@ -168,7 +170,18 @@ class TunableOp {
     // Do nothing if we are not playing around with params
   }
 
-  virtual ~TunableOp() = default;
+  std::string Signature() {
+    // According to C++17 standard https://wg21.link/n4659 section 15.7.4
+    // > if the operand of typeid refers to the
+    // > object under construction or destruction, typeid yields the std::type_info object representing the constructor
+    // > or destructor’s class.
+    // So delay the op signature generation. See https://github.com/microsoft/onnxruntime/pull/14709
+    std::lock_guard<OrtMutex> g(signature_mutex_);
+    if (signature_.empty()) {
```

Comment:
this operation is called every iteration. so it's not good to add mutex here.
should move mutex into if statement, and after getting lock, check signature again.
code like this:

```
if (signature.empty()) {
   std::lock_gurad<OrtMutex> g(signature_mutex_);
   if (signature.empty()) {
    siganature_ = CreateSignature();
   }
}
```

Reply:
Then it will be a misused of lock.

Reply:
Not really. It's a standard test-and-test-and-set pattern that is being proposed: https://en.wikipedia.org/wiki/Test_and_test-and-set

Reply:
@abudup Don't use this pattern. It is not safe at all. It has a formal name call double-checked lock pattern. Sutter has an article on how it can be unsafe unless you implement it directly within your compiler. I will switch to call_once here.